REGRESSION(278568@main): UI process crash when event replies are received during process termination

charliewolfe opened the pull request:
#### 4f158fce2282ed8ee934e847cfade3bc7621969f
<pre>
REGRESSION(278568@main): UI process crash when event replies are received during process termination
<a href="https://bugs.webkit.org/show_bug.cgi?id=274688">https://bugs.webkit.org/show_bug.cgi?id=274688</a>
<a href="https://rdar.apple.com/128236488">rdar://128236488</a>

Reviewed by Chris Dumez.

After 278568@main, if mouse/key event replies are received after web process termination has begun but
its `WebProcessProxy` has not been destructed, we try to dequeue from an empty queue. We should not try
to dequeue events if the current process is being terminated, because the events queues will have already
been cleared.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent):
* Tools/TestWebKitAPI/Tests/mac/KeyboardEventTests.mm:
(TestWebKitAPI::TEST(KeyboardEventTests, TerminateWebContentProcessDuringKeyEventHandling)):
* Tools/TestWebKitAPI/Tests/mac/MouseEventTests.mm:
(TestWebKitAPI::TEST(MouseEventTests, TerminateWebContentProcessDuringMouseEventHandling)):

Canonical link: <a href="https://commits.webkit.org/279333@main">https://commits.webkit.org/279333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/955a396c728ea17879bd6c3aeb1d4ec3555c9078

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56447 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55473 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39599 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3624 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43098 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2522 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30409 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45899 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24228 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27392 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3227 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2050 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49198 "Found 1 new API test failure: TestWebKitAPI.SleepDisabler.Crash (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58043 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3342 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50500 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29529 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46122 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49812 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30449 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7816 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29285 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Tools/TestWebKitAPI/Tests/mac/KeyboardEventTests.mm:
```diff
@@ -126,6 +128,21 @@ static void arrowKeyDownWithKeyRepeat(WKWebView *webView, unsigned keyRepeatCoun
 
+TEST(KeyboardEventTests, TerminateWebContentProcessDuringKeyEventHandling)
+{
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600)]);
```
`auto` -> `RetainPtr`

charliewolfe committed:
REGRESSION(278568@main): UI process crash when event replies are received during process termination
https://bugs.webkit.org/show_bug.cgi?id=274688
rdar://128236488

Reviewed by Chris Dumez.

After 278568@main, if mouse/key event replies are received after web process termination has begun but
its `WebProcessProxy` has not been destructed, we try to dequeue from an empty queue. We should not try
to dequeue events if the current process is being terminated, because the events queues will have already
been cleared.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent):
* Tools/TestWebKitAPI/Tests/mac/KeyboardEventTests.mm:
(TestWebKitAPI::TEST(KeyboardEventTests, TerminateWebContentProcessDuringKeyEventHandling)):
* Tools/TestWebKitAPI/Tests/mac/MouseEventTests.mm:
(TestWebKitAPI::TEST(MouseEventTests, TerminateWebContentProcessDuringMouseEventHandling)):

Canonical link: https://commits.webkit.org/279333@main
